index.rst

```diff
@@ -1,61 +1,173 @@
+
+Existing Requirements
+---------------------
+
+The database is referenced in a handful of existing project requirements documents:
+
+ - OSS-REQ-0128 "Alerts" :cite:`LSE-30`:
+
+     The Level 1 Data Products shall include the Alerts produced as part  of the nightly Alert Production.
+
+ - OSS-REQ-0185 "Transient Alert Query" :cite:`LSE-30`:
+
+     All published transient alerts, as well as all reprocessed historical alerts generated as part of a Data Release, shall be available for query.
+
+ - DMS-REQ-0094 "Keep Historical Alert Archive" :cite:`LSE-61`:
 
-.. TODO: Delete the note below before merging new content to the master branch.
+     The DMS shall preserve and keep in an accessible state an alert archive with all issued alerts for a historical record and for false alert analysis.
+
+ - DMS-PRTL-REQ-0033 "Queries on the Alerts Database" :cite:`LDM-554`:
+
+     The Portal aspect shall provide a query interface to the Alert Database, allowing searches based on parameters which shall include, but may not be limited to, Alert ID, time of alert, position on the sky, filter, and alert characteristics.
+
+ - DMS-PRTL-REQ-0048 "Alert Visualization" :cite:`LDM-554`:
+
+     The Portal aspect shall provide for the users a "property sheet" for the contents of an alert packet including, but not necessarily limited to, the alert postage stamp image, the postage stamp time series, the photometric time series, the source and object information (e.g., position, brightness).
+
+Proposed Implementation
+=======================
+
+We can satisfy these design inputs by storing serialized Avro alert data (the same bytes sent via Kafka to brokers) in a S3-like object store, indexed by a unique alert ID.
+Each alert packet corresponds to one object in the object store.
 
 .. note::
 
-   **This technote is not yet published.**
+   An alternative would be to combine many packets into a block in the object store, perhaps of about 100 alert packets.
+   This might permit more efficient storage and retrieval.
+   Storage might be more efficient because compression would be better when storing many alerts.
+   Retrieval might be more efficient because it might tune the outgoing flows into a smaller number of TCP connections which get a chance to grow window sizes.
 
-   Design document for a database of all alerts sent by the Rubin observatory.
+   But this would be more complex, and make writing more difficult, as writes need to append to existing data which would require coordination between writers, so this design sticks to a simpler structure.
+
+An object store is used because it is cheap, scales well to handle terabytes of data, and should support parallel retrieval reasonably well.
+
+Writing data
+------------
+
+When the alert production pipeline has computed a new alert packet, it should be careful to write it to the alert archive before publishing to a Kafka topic, to ensure that it gets archived. The overall flow should be:
```

Comment:
If that latency to the archive is too hard to guarantee, I wonder if temporary staging to a local filesystem would be sufficient.

Reply:
~sub-second median write times is almost certainly easy. ~sub-second p99.9 write times is almost certainly extremely difficult and would require careful TCP tuning. (The same is true of the writes to Kafka, by the way).

I don't think hard time limits are compatible with hard data durability requirements. There's probably a formal way to state that (maybe [PACELC?](https://brooker.co.za/blog/2015/09/26/cap-durability.html)). I think we have to choose one or the other.

Temporary staging to a local filesystem doesn't seem to guarantee durability to me; I think it pushes the problem into an additional system (if you write to Kafka after the filesystem write, then you're now depending upon the filesystem -> DB flow working flawlessly, which is not realistic).

Reply:
@ktlim's idea of recovering alerts from downstream brokers is good, and it suggests a further change. We could write to AlertDB from a consumer hanging off of the alerts Kafka topic.

This means we are writing to the alert database after we've published alerts, but Kafka could keep a healthy buffer (7 days, or something) so we can be confident _enough_ that we'd preserve the alert in a database.

This would add latency to the writes to the Alert DB, but that seems acceptable to me - very low latency is not one of the driving goals of the thing.

Reply:
I've revised to say that we should run a Kafka consumer which copies data into the alert database.